stock_storage_type_buffer/models/stock_location_storage_buffer.py

```diff
@@ -0,0 +1,88 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+
+from odoo import _, api, fields, models
+
+
+class StockLocationStorageBuffer(models.Model):
+    _name = "stock.location.storage.buffer"
+    _description = "Location Storage Buffer"
+
+    MAX_LOCATIONS_IN_NAME_GET = 3
+    MAX_LOCATIONS_IN_HELP = 10
+
+    buffer_location_ids = fields.Many2many(
+        comodel_name="stock.location",
+        relation="stock_location_storage_buffer_stock_location_buffer_rel",
+        required=True,
+        help="Buffers where goods are temporarily stored. When all these "
+        "locations contain goods or will receive goods,"
+        " the destination locations are not available for putaway.",
+    )
+    location_ids = fields.Many2many(
+        comodel_name="stock.location",
+        relation="stock_location_storage_buffer_stock_location_rel",
+        required=True,
+        help="Destination locations (including sublocations) that will be"
+        " unreachable for putaway if the buffers are full.",
+    )
+    help_message = fields.Html(compute="_compute_help_message")
+    active = fields.Boolean(default=True)
+
+    @api.depends("buffer_location_ids", "location_ids")
+    def _compute_help_message(self):
+        """Generate a description of the storage buffer for humans"""
+        for record in self:
+            if not (record.buffer_location_ids and record.location_ids):
+                record.help_message = _(
+                    "<p>Select buffer locations and locations "
+                    "blocked for putaways when the buffer locations "
+                    "already contain goods or have move lines "
+                    "reaching them.</p>"
+                )
+                continue
+
+            record.help_message = record._help_message()
+
+    def buffers_have_capacity(self):
+        self.ensure_one()
+        buffer_locations = self.buffer_location_ids.leaf_location_ids
+        return any(location.location_is_empty for location in buffer_locations)
```

Comment:
The buffer should also be considered empty as soon as
sum(location.out_move_lines_ids.qty_done) = sum(location.quant_ids.mapped("quantity"))
Probably to put in _compute_location_is_empty.
So when someone takes what is on the buffer by writing the done quantity on the move line, the buffer is physically empty and we should be able to put again something on it.
This detail is quite important for the selection of the right set of allowed locations.

Reply:
@jbaudoux can you check if this is what you expect: https://github.com/OCA/wms/pull/58/commits/07b342ec16183c59e0eab6f9d7e157917d92658f ?

Reply:
Perfect. Thanks